[SUPPORT] NPE when reading from a CDC-enabled table with null values

**Describe the problem you faced**

Reading from a CDC-enabled Hudi table results in a cryptic NPE error as shown in the stacktrace section. The error occurs after writing to the table with `WriteOperationType.BULK_INSERT`. Writing with `UPSERT` does not cause any issues when reading.

**To Reproduce**

Steps to reproduce the behavior:

1. Write to Hudi table:
```
ds.write()
                .format("hudi")
                //... some options omitted for brevity
                .option(DataSourceWriteOptions.OPERATION().key(), WriteOperationType.BULK_INSERT.name())
                .option(HoodieTableConfig.NAME.key(), "foo")
                .option(HoodieTableConfig.CDC_ENABLED.key(), "true")
                .option(HoodieTableConfig.CDC_SUPPLEMENTAL_LOGGING_MODE.key(), HoodieCDCSupplementalLoggingMode.data_before_after.name())
                .mode(Overwrite)
                .save(...);
```
2. Read from the table
```
 spark.readStream()
                    .format("hudi")
                    .option("hoodie.datasource.query.incremental.format", "cdc")
                    .option("hoodie.datasource.query.type", "incremental")
                    .load(...)
                    .writeStream()
                    .foreachBatch((batchDF, batchId) -> {
                        batchDF.show(); // FIXME: this line results in an NPE 
                    }).start().awaitTermination();
```

**Expected behavior**

Expecting same results when reading from the table regardless of the write operation used. Expecting "Support all the write operations" as stated in [the Design Goals section of the CDC RFC](https://github.com/apache/hudi/blob/master/rfc/rfc-51/rfc-51.md#design-goals).

**Environment Description**

* Hudi version : 0.13.1

* Spark version : 3.3.2

* Hive version : n/a

* Hadoop version : 3.3.3

* Storage (HDFS/S3/GCS..) : local file system for testing, S3 in the long run

* Running on Docker? (yes/no) : yes


**Stacktrace**

```23/06/21 22:10:33 ERROR TaskSetManager: Task 0 in stage 0.0 failed 4 times; aborting job
23/06/21 22:10:33 ERROR MicroBatchExecution: Query [id = e4cb7dc2-dba3-403e-8d9d-d5491dc7e8b0, runId = 1837bba4-981a-4edb-be4d-9d877abf65bd] terminated with error
org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 0.0 failed 4 times, most recent failure: Lost task 0.3 in stage 0.0 (TID 3) (172.19.0.3 executor 0): java.lang.NullPointerException: Cannot invoke "org.apache.spark.unsafe.types.UTF8String.toString()" because the return value of "org.apache.spark.sql.catalyst.InternalRow.getUTF8String(int)" is null
	at org.apache.spark.sql.catalyst.InternalRow.getString(InternalRow.scala:34)
	at org.apache.hudi.cdc.HoodieCDCRDD$CDCFileGroupIterator.$anonfun$convertRowToJsonString$1(HoodieCDCRDD.scala:562)
	at org.apache.hudi.cdc.HoodieCDCRDD$CDCFileGroupIterator.$anonfun$convertRowToJsonString$1$adapted(HoodieCDCRDD.scala:559)
	at scala.collection.immutable.List.foreach(List.scala:431)
	at org.apache.hudi.cdc.HoodieCDCRDD$CDCFileGroupIterator.convertRowToJsonString(HoodieCDCRDD.scala:559)
	at org.apache.hudi.cdc.HoodieCDCRDD$CDCFileGroupIterator.loadNext(HoodieCDCRDD.scala:295)
	at org.apache.hudi.cdc.HoodieCDCRDD$CDCFileGroupIterator.hasNextInternal(HoodieCDCRDD.scala:263)
	at org.apache.hudi.cdc.HoodieCDCRDD$CDCFileGroupIterator.hasNext(HoodieCDCRDD.scala:284)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:760)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:760)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$getByteArrayRdd$1(SparkPlan.scala:364)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:890)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:890)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:365)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:329)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:136)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:548)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1504)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:551)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2672)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2608)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2607)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2607)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1182)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1182)
	at scala.Option.foreach(Option.scala:407)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1182)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2860)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2802)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2791)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:952)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2238)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2259)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2278)
	at org.apache.spark.sql.execution.SparkPlan.executeTake(SparkPlan.scala:506)
	at org.apache.spark.sql.execution.SparkPlan.executeTake(SparkPlan.scala:459)
	at org.apache.spark.sql.execution.CollectLimitExec.executeCollect(limit.scala:48)
	at org.apache.spark.sql.Dataset.collectFromPlan(Dataset.scala:3868)
	at org.apache.spark.sql.Dataset.$anonfun$head$1(Dataset.scala:2863)
	at org.apache.spark.sql.Dataset.$anonfun$withAction$2(Dataset.scala:3858)
	at org.apache.spark.sql.execution.QueryExecution$.withInternalError(QueryExecution.scala:510)
	at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:3856)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$6(SQLExecution.scala:109)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:169)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:95)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
	at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3856)
	at org.apache.spark.sql.Dataset.head(Dataset.scala:2863)
	at org.apache.spark.sql.Dataset.take(Dataset.scala:3084)
	at org.apache.spark.sql.Dataset.getRows(Dataset.scala:288)
	at org.apache.spark.sql.Dataset.showString(Dataset.scala:327)
	at org.apache.spark.sql.Dataset.show(Dataset.scala:808)
	at org.apache.spark.sql.Dataset.show(Dataset.scala:767)
	at org.apache.spark.sql.Dataset.show(Dataset.scala:776)
	...
Caused by: java.lang.NullPointerException
	at org.apache.spark.sql.catalyst.InternalRow.getString(InternalRow.scala:34)
	at org.apache.hudi.cdc.HoodieCDCRDD$CDCFileGroupIterator.$anonfun$convertRowToJsonString$1(HoodieCDCRDD.scala:562)
	at org.apache.hudi.cdc.HoodieCDCRDD$CDCFileGroupIterator.$anonfun$convertRowToJsonString$1$adapted(HoodieCDCRDD.scala:559)
	at scala.collection.immutable.List.foreach(List.scala:431)
	at org.apache.hudi.cdc.HoodieCDCRDD$CDCFileGroupIterator.convertRowToJsonString(HoodieCDCRDD.scala:559)
	at org.apache.hudi.cdc.HoodieCDCRDD$CDCFileGroupIterator.loadNext(HoodieCDCRDD.scala:295)
	at org.apache.hudi.cdc.HoodieCDCRDD$CDCFileGroupIterator.hasNextInternal(HoodieCDCRDD.scala:263)
	at org.apache.hudi.cdc.HoodieCDCRDD$CDCFileGroupIterator.hasNext(HoodieCDCRDD.scala:284)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:760)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:760)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$getByteArrayRdd$1(SparkPlan.scala:364)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:890)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:890)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:365)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:329)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:136)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:548)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1504)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:551)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

--- a/hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/cdc/HoodieCDCRDD.scala
+++ b/hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/cdc/HoodieCDCRDD.scala
@@ -561,7 +561,7 @@ class HoodieCDCRDD(
       originTableSchema.structTypeSchema.zipWithIndex.foreach {
         case (field, idx) =>
           if (field.dataType.isInstanceOf[StringType]) {
-            map(field.name) = record.getString(idx)
+            map(field.name) = Option(record.getUTF8String(idx)).map(_.toString).orNull
           } else {
             map(field.name) = record.get(idx, field.dataType)
           }

--- a/hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/cdc/HoodieCDCRDD.scala
+++ b/hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/cdc/HoodieCDCRDD.scala
@@ -18,9 +18,6 @@
 
 package org.apache.hudi.cdc
 
-import com.fasterxml.jackson.annotation.JsonInclude.Include
-import com.fasterxml.jackson.databind.{DeserializationFeature, ObjectMapper}
-import com.fasterxml.jackson.module.scala.DefaultScalaModule
 import org.apache.avro.Schema
 import org.apache.avro.generic.{GenericData, GenericRecord, IndexedRecord}
 import org.apache.hadoop.fs.Path
@@ -47,7 +44,7 @@ import org.apache.spark.sql.avro.HoodieAvroDeserializer
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.expressions.Projection
 import org.apache.spark.sql.execution.datasources.PartitionedFile
-import org.apache.spark.sql.types.{StringType, StructType}
+import org.apache.spark.sql.types.StructType
 import org.apache.spark.unsafe.types.UTF8String
 import org.apache.spark.{Partition, SerializableWritable, TaskContext}
 
@@ -155,14 +152,6 @@ class HoodieCDCRDD(
         recordMergerImpls = List(classOf[HoodieAvroRecordMerger].getName),
         recordMergerStrategy = HoodieRecordMerger.DEFAULT_MERGER_STRATEGY_UUID
       )
-    }
-
-    private lazy val mapper: ObjectMapper = {
-      val _mapper = new ObjectMapper
-      _mapper.setSerializationInclusion(Include.NON_ABSENT)
-      _mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
-      _mapper.registerModule(DefaultScalaModule)
-      _mapper
     }
 
     protected override val avroSchema: Schema = new Schema.Parser().parse(originTableSchema.avroSchemaStr)
@@ -557,16 +546,7 @@ class HoodieCDCRDD(
      * Convert InternalRow to json string.
      */
     private def convertRowToJsonString(record: InternalRow): UTF8String = {
-      val map = scala.collection.mutable.Map.empty[String, Any]
-      originTableSchema.structTypeSchema.zipWithIndex.foreach {
-        case (field, idx) =>
-          if (field.dataType.isInstanceOf[StringType]) {
-            map(field.name) = Option(record.getUTF8String(idx)).map(_.toString).orNull
-          } else {
-            map(field.name) = record.get(idx, field.dataType)
-          }
-      }
-      convertToUTF8String(mapper.writeValueAsString(map))
+      new InternalRowToJsonStringConverter(originTableSchema).convertRowToJsonString(record)
     }
 
     /**

--- a/hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/cdc/HoodieCDCRDD.scala
+++ b/hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/cdc/HoodieCDCRDD.scala
@@ -234,6 +234,8 @@ class HoodieCDCRDD(
      */
     private var afterImageRecords: mutable.Map[String, InternalRow] = mutable.Map.empty
 
+    private var internalRowToJsonStringConverter = new InternalRowToJsonStringConverter(originTableSchema)
+
     private def needLoadNextFile: Boolean = {
       !recordIter.hasNext &&
         !logRecordIter.hasNext &&
@@ -546,7 +548,7 @@ class HoodieCDCRDD(
      * Convert InternalRow to json string.
      */
     private def convertRowToJsonString(record: InternalRow): UTF8String = {
-      new InternalRowToJsonStringConverter(originTableSchema).convertRowToJsonString(record)
+      internalRowToJsonStringConverter.convert(record)
     }
 
     /**

--- a/hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/cdc/InternalRowToJsonStringConverter.scala
+++ b/hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/cdc/InternalRowToJsonStringConverter.scala
@@ -36,7 +36,7 @@ class InternalRowToJsonStringConverter(originTableSchema: HoodieTableSchema) {
     _mapper
   }
 
-  def convertRowToJsonString(record: InternalRow): UTF8String = {
+  def convert(record: InternalRow): UTF8String = {
     val map = scala.collection.mutable.Map.empty[String, Any]
     originTableSchema.structTypeSchema.zipWithIndex.foreach {
       case (field, idx) =>

--- a/hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/cdc/InternalRowToJsonStringConverter.scala
+++ b/hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/cdc/InternalRowToJsonStringConverter.scala
@@ -1,3 +1,21 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 package org.apache.hudi.cdc
 
 import com.fasterxml.jackson.annotation.JsonInclude.Include

--- a/hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/cdc/InternalRowToJsonStringConverter.scala
+++ b/hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/cdc/InternalRowToJsonStringConverter.scala
@@ -1,0 +1,33 @@
+package org.apache.hudi.cdc
+
+import com.fasterxml.jackson.annotation.JsonInclude.Include
+import com.fasterxml.jackson.databind.{DeserializationFeature, ObjectMapper}
+import com.fasterxml.jackson.module.scala.DefaultScalaModule
+import org.apache.hudi.HoodieTableSchema
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.types.StringType
+import org.apache.spark.unsafe.types.UTF8String
+
+class InternalRowToJsonStringConverter(originTableSchema: HoodieTableSchema) {
+
+  private lazy val mapper: ObjectMapper = {
+    val _mapper = new ObjectMapper
+    _mapper.setSerializationInclusion(Include.NON_ABSENT)
+    _mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+    _mapper.registerModule(DefaultScalaModule)
+    _mapper
+  }
+
+  def convertRowToJsonString(record: InternalRow): UTF8String = {
+    val map = scala.collection.mutable.Map.empty[String, Any]
+    originTableSchema.structTypeSchema.zipWithIndex.foreach {
+      case (field, idx) =>
+        if (field.dataType.isInstanceOf[StringType]) {
+          map(field.name) = Option(record.getUTF8String(idx)).map(_.toString).orNull
+        } else {
+          map(field.name) = record.get(idx, field.dataType)
+        }
+    }
+    UTF8String.fromString(mapper.writeValueAsString(map))
+  }
+}

--- a/hudi-spark-datasource/hudi-spark-common/src/test/scala/org/apache/hudi/cdc/TestInternalRowToJsonStringConverter.scala
+++ b/hudi-spark-datasource/hudi-spark-common/src/test/scala/org/apache/hudi/cdc/TestInternalRowToJsonStringConverter.scala
@@ -1,0 +1,56 @@
+package org.apache.hudi.cdc
+
+import org.apache.hudi.HoodieTableSchema
+import org.apache.hudi.internal.schema.InternalSchema
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.types.{DataTypes, Metadata, StructField, StructType}
+import org.apache.spark.unsafe.types.UTF8String
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Test
+
+class TestInternalRowToJsonStringConverter {
+  private val converter = new InternalRowToJsonStringConverter(hoodieTableSchema)
+
+  @Test
+  def emptyRow(): Unit = {
+    val converter = new InternalRowToJsonStringConverter(emptyHoodieTableSchema)
+    val row = InternalRow.empty
+    val converted = converter.convertRowToJsonString(row)
+    assertEquals("{}", converted.toString)
+  }
+
+  @Test
+  def nonEmptyRow(): Unit = {
+    val row = InternalRow.fromSeq(Seq(1, UTF8String.fromString("foo")))
+    val converted = converter.convertRowToJsonString(row)
+    assertEquals("{\"name\":\"foo\",\"uuid\":1}", converted.toString)
+  }
+
+  @Test
+  def emptyString(): Unit = {
+    val row = InternalRow.fromSeq(Seq(1, UTF8String.EMPTY_UTF8))
+    val converted = converter.convertRowToJsonString(row)
+    assertEquals("{\"name\":\"\",\"uuid\":1}", converted.toString)
+  }
+
+  @Test
+  def nullString(): Unit = {
+    val row = InternalRow.fromSeq(Seq(1, null))
+    val converted = converter.convertRowToJsonString(row)
+    assertEquals("{\"uuid\":1}", converted.toString)
+  }
+
+  private def hoodieTableSchema: HoodieTableSchema = {
+    val structTypeSchema = new StructType(Array[StructField](
+      StructField("uuid", DataTypes.IntegerType, nullable = false, Metadata.empty),
+      StructField("name", DataTypes.StringType, nullable = true, Metadata.empty)))
+    val avroSchemaStr: String = "{\"type\": \"record\",\"name\": \"test\",\"fields\": [{\"name\": \"uuid\",\"type\": \"int\"},{\"name\": \"name\",\"type\": \"string\"}]}"
+    HoodieTableSchema(structTypeSchema, avroSchemaStr, Option.empty[InternalSchema])
+  }
+
+  private def emptyHoodieTableSchema: HoodieTableSchema = {
+    val structTypeSchema = new StructType()
+    val avroSchemaStr = "{\"type\": \"record\",\"name\": \"test\",\"fields\": []}"
+    HoodieTableSchema(structTypeSchema, avroSchemaStr, Option.empty[InternalSchema])
+  }
+}

--- a/hudi-spark-datasource/hudi-spark-common/src/test/scala/org/apache/hudi/cdc/TestInternalRowToJsonStringConverter.scala
+++ b/hudi-spark-datasource/hudi-spark-common/src/test/scala/org/apache/hudi/cdc/TestInternalRowToJsonStringConverter.scala
@@ -41,34 +41,38 @@ class TestInternalRowToJsonStringConverter {
   def nonEmptyRow(): Unit = {
     val row = InternalRow.fromSeq(Seq(1, UTF8String.fromString("foo")))
     val converted = converter.convertRowToJsonString(row)
-    assertEquals("{\"name\":\"foo\",\"uuid\":1}", converted.toString)
+    assertEquals("""{"name":"foo","uuid":1}""", converted.toString)
   }
 
   @Test
   def emptyString(): Unit = {
     val row = InternalRow.fromSeq(Seq(1, UTF8String.EMPTY_UTF8))
     val converted = converter.convertRowToJsonString(row)
-    assertEquals("{\"name\":\"\",\"uuid\":1}", converted.toString)
+    assertEquals("""{"name":"","uuid":1}""", converted.toString)
   }
 
   @Test
   def nullString(): Unit = {
     val row = InternalRow.fromSeq(Seq(1, null))
     val converted = converter.convertRowToJsonString(row)
-    assertEquals("{\"uuid\":1}", converted.toString)
+    assertEquals("""{"uuid":1}""", converted.toString)
   }
 
   private def hoodieTableSchema: HoodieTableSchema = {
     val structTypeSchema = new StructType(Array[StructField](
       StructField("uuid", DataTypes.IntegerType, nullable = false, Metadata.empty),
       StructField("name", DataTypes.StringType, nullable = true, Metadata.empty)))
-    val avroSchemaStr: String = "{\"type\": \"record\",\"name\": \"test\",\"fields\": [{\"name\": \"uuid\",\"type\": \"int\"},{\"name\": \"name\",\"type\": \"string\"}]}"
+    val avroSchemaStr: String =
+      """{"type": "record", "name": "test", "fields": [
+        |{"name": "uuid", "type": "int"},
+        |{"name": "name", "type": "string"}
+        |]}""".stripMargin
     HoodieTableSchema(structTypeSchema, avroSchemaStr, Option.empty[InternalSchema])
   }
 
   private def emptyHoodieTableSchema: HoodieTableSchema = {
     val structTypeSchema = new StructType()
-    val avroSchemaStr = "{\"type\": \"record\",\"name\": \"test\",\"fields\": []}"
+    val avroSchemaStr = """{"type": "record", "name": "test", "fields": []}"""
     HoodieTableSchema(structTypeSchema, avroSchemaStr, Option.empty[InternalSchema])
   }
 }

--- a/hudi-spark-datasource/hudi-spark-common/src/test/scala/org/apache/hudi/cdc/TestInternalRowToJsonStringConverter.scala
+++ b/hudi-spark-datasource/hudi-spark-common/src/test/scala/org/apache/hudi/cdc/TestInternalRowToJsonStringConverter.scala
@@ -33,28 +33,28 @@ class TestInternalRowToJsonStringConverter {
   def emptyRow(): Unit = {
     val converter = new InternalRowToJsonStringConverter(emptyHoodieTableSchema)
     val row = InternalRow.empty
-    val converted = converter.convertRowToJsonString(row)
+    val converted = converter.convert(row)
     assertEquals("{}", converted.toString)
   }
 
   @Test
   def nonEmptyRow(): Unit = {
     val row = InternalRow.fromSeq(Seq(1, UTF8String.fromString("foo")))
-    val converted = converter.convertRowToJsonString(row)
+    val converted = converter.convert(row)
     assertEquals("""{"name":"foo","uuid":1}""", converted.toString)
   }
 
   @Test
   def emptyString(): Unit = {
     val row = InternalRow.fromSeq(Seq(1, UTF8String.EMPTY_UTF8))
-    val converted = converter.convertRowToJsonString(row)
+    val converted = converter.convert(row)
     assertEquals("""{"name":"","uuid":1}""", converted.toString)
   }
 
   @Test
   def nullString(): Unit = {
     val row = InternalRow.fromSeq(Seq(1, null))
-    val converted = converter.convertRowToJsonString(row)
+    val converted = converter.convert(row)
     assertEquals("""{"uuid":1}""", converted.toString)
   }
 

--- a/hudi-spark-datasource/hudi-spark-common/src/test/scala/org/apache/hudi/cdc/TestInternalRowToJsonStringConverter.scala
+++ b/hudi-spark-datasource/hudi-spark-common/src/test/scala/org/apache/hudi/cdc/TestInternalRowToJsonStringConverter.scala
@@ -1,3 +1,21 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 package org.apache.hudi.cdc
 
 import org.apache.hudi.HoodieTableSchema

--- a/hudi-spark-datasource/hudi-spark-common/src/test/scala/org/apache/hudi/cdc/TestInternalRowToJsonStringConverter.scala
+++ b/hudi-spark-datasource/hudi-spark-common/src/test/scala/org/apache/hudi/cdc/TestInternalRowToJsonStringConverter.scala
@@ -23,7 +23,7 @@ import org.apache.hudi.internal.schema.InternalSchema
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.types.{DataTypes, Metadata, StructField, StructType}
 import org.apache.spark.unsafe.types.UTF8String
-import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Assertions.{assertEquals, assertTrue}
 import org.junit.jupiter.api.Test
 
 class TestInternalRowToJsonStringConverter {
@@ -55,7 +55,7 @@ class TestInternalRowToJsonStringConverter {
   def nullString(): Unit = {
     val row = InternalRow.fromSeq(Seq(1, null))
     val converted = converter.convert(row)
-    assertEquals("""{"uuid":1}""", converted.toString)
+    assertTrue(converted.toString.equals("""{"uuid":1}""") || converted.toString.equals("""{"name":null,"uuid":1}"""))
   }
 
   private def hoodieTableSchema: HoodieTableSchema = {

